plugin/PlayAssetDelivery/src/main/java/com/google/play/core/godot/assetpacks/utils/PlayAssetDeliveryUtils.java

```
@@ -0,0 +1,90 @@
+ */
+public class PlayAssetDeliveryUtils {
+
+  private static final String TAG = "PlayAssetDeliveryUtils";
+
+  public static Dictionary convertAssetPackStateToDictionary(AssetPackState assetPackState) {
+    try {
+      Dictionary returnDict = new Dictionary();
+      returnDict.put("bytesDownloaded", assetPackState.bytesDownloaded());
+      returnDict.put("errorCode", assetPackState.errorCode());
+      returnDict.put("name", assetPackState.name());
+      returnDict.put("status", assetPackState.status());
+      returnDict.put("totalBytesToDownload", assetPackState.totalBytesToDownload());
+      returnDict.put("transferProgressPercentage", assetPackState.transferProgressPercentage());
+      return returnDict;
+    } catch (Exception e) {
+      Log.w(TAG, "Exception while converting AssetPackState object to Dictionary!", e);
+      return null;
+    }
+  }
+
+  public static AssetPackState convertDictionaryToAssetPackState(final Dictionary dict) {
+    try {
+      AssetPackState packState =
+          new AssetPackState() {
+            @Override
+            public String name() {
+              return (String) dict.get("name");
+            }
+
+            @Override
+            public int status() {
+              return (int) dict.get("status");
+            }
+
+            @Override
+            public int errorCode() {
+              return (int) dict.get("errorCode");
+            }
+
+            @Override
+            public long bytesDownloaded() {
+              return (long) dict.get("bytesDownloaded");
+            }
+
+            @Override
+            public long totalBytesToDownload() {
+              return (long) dict.get("totalBytesToDownload");
+            }
+
+            @Override
+            public int transferProgressPercentage() {
+              return (int) dict.get("transferProgressPercentage");
+            }
+          };
+      return packState;
+    } catch (Exception e) {
```

Would this exception ever throw? It looks like the resultant `AssetPackState` only throws an exception when one of its methods is called. It would be better to have `convertDictionaryToAssetPackState` throw immediately if the provided dictionary is invalid, rather than create an object that might be invalid, but the caller won't know until they try to access it.

It might be good to replace this anonymous class with a real class that extends AssetPackState, maybe `AssetPackStateFromDictionary` or something. Then you could assign all these values at constructor time without needing to refer to the dictionary.

---

Yup that sounds like a good idea. It will also make the util class look cleaner. Let me try it real quick.

---

Replaced anonymous class with AssetPackStateFromDictionary class. Since the constructor throws NullPointerException and ClassCastException, we need to wrap it inside try catch block.

---

Should we throw an `IllegalArgumentException` exception rather than return null?